fix: Validate SetStatisticsUpdate correctly

ragnard opened the pull request:
Closes #2865

Previously the pydantic `@model_validator` on `SetStatisticsUpdate` would fail because it assumed statistics was a model instance. In a "before"" validator that is not necessarily case.

Check type and handle both models and dicts instead.

### Before

```
>>> import pyiceberg.table.update
>>> pyiceberg.table.update.SetStatisticsUpdate.model_validate({'statistics': {'snapshot-id': 1234, 'file-size-in-bytes': 0, 'statistics-path': '', 'file-footer-size-in-bytes': 0, 'blob-metadata': []}})
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    pyiceberg.table.update.SetStatisticsUpdate.model_validate({'statistics': {'snapshot-id': 1234, 'file-size-in-bytes': 0, 'statistics-path': '', 'file-footer-size-in-bytes': 0, 'blob-metadata': []}})
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ragge/projects/github.com/ragnard/iceberg-python/.venv/lib/python3.14/site-packages/pydantic/main.py", line 716, in model_validate
    return cls.__pydantic_validator__.validate_python(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        obj,
        ^^^^
    ...<5 lines>...
        by_name=by_name,
        ^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/ragge/projects/github.com/ragnard/iceberg-python/pyiceberg/table/update/__init__.py", line 191, in validate_snapshot_id
    data["snapshot_id"] = stats.snapshot_id
                          ^^^^^^^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'snapshot_id'
 ```

### After

```
>>> import pyiceberg.table.update
>>> pyiceberg.table.update.SetStatisticsUpdate.model_validate({'statistics': {'snapshot-id': 1234, 'file-size-in-bytes': 0, 'statistics-path': '', 'file-footer-size-in-bytes': 0, 'blob-metadata': []}})
SetStatisticsUpdate(action='set-statistics', statistics=StatisticsFile(snapshot_id=1234, statistics_path='', file_size_in_bytes=0, file_footer_size_in_bytes=0, key_metadata=None, blob_metadata=[]), snapshot_id=1234)
``` 


# Rationale for this change

## Are these changes tested?

Yes, but only using the two-liners above.

## Are there any user-facing changes?

No.

<!-- In the case of user-facing changes, please add the changelog label. -->

kevinjqliu reviewed (changes requested):
oops didnt mean to approve

ragnard commented:
@kevinjqliu Ok, do you want me to change the fix so that `snapshot_id` is still there, but just not automatically populated?

kevinjqliu commented on pyiceberg/table/update/__init__.py:
add a test, something like:

```
def test_set_statistics_update_without_snapshot_id(table_v2_with_statistics: Table) -> None:
    """Test that SetStatisticsUpdate auto-populates snapshot_id from statistics."""
    snapshot_id = table_v2_with_statistics.metadata.current_snapshot_id

    blob_metadata = BlobMetadata(
        type="apache-datasketches-theta-v1",
        snapshot_id=snapshot_id,
        sequence_number=2,
        fields=[1],
        properties={"prop-key": "prop-value"},
    )

    statistics_file = StatisticsFile(
        snapshot_id=snapshot_id,
        statistics_path="s3://bucket/warehouse/stats.puffin",
        file_size_in_bytes=124,
        file_footer_size_in_bytes=27,
        blob_metadata=[blob_metadata],
    )

    # Create update without explicitly providing snapshot_id
    update = SetStatisticsUpdate(statistics=statistics_file)

    # Validator should auto-populate snapshot_id from statistics
    assert update.snapshot_id == snapshot_id

    new_metadata = update_table_metadata(
        table_v2_with_statistics.metadata,
        (update,),
    )

    assert len(new_metadata.statistics) == 2
    updated_statistics = [stat for stat in new_metadata.statistics if stat.snapshot_id == snapshot_id]
    assert len(updated_statistics) == 1
```

kevinjqliu reviewed (commented):
Looks like the java implementation has not deprecated the top-level `snapshot_id`. lets proceed with this change since it improves the current validation logic. 

Thanks for the PR! Lets add a test and it should be good to go!

ragnard commented:
> Looks like the java implementation has not deprecated the top-level `snapshot_id`. lets proceed with this change since it improves the current validation logic.
> 
> Thanks for the PR! Lets add a test and it should be good to go!

@kevinjqliu Thanks for the (quick!) review. I've changed the fix a bit:

1) Since the model is frozen it is not possible to use direct assignment, but it is also not possible to use `model_copy` like I did because an "after" validator needs to return the same model instance. I've reverted to a "before" validator, but handle the `dict` case properly.

2) For testing, the issue is not really about tables with statistics, but that it was not possible to instantiate the `SetStatisticsUpdate` model from a dict. I added an `model_roundtrips` helper that is now used to check that any model roundtrips (`model -> dict -> model`) correctly.

Please let me know if you want further changes.

kevinjqliu commented on pyiceberg/table/update/__init__.py:
```diff
@@ -181,9 +181,15 @@ class SetStatisticsUpdate(IcebergBaseModel):
+        if isinstance(stats, StatisticsFile):
+            snapshot_id = stats.snapshot_id
+        elif isinstance(stats, dict):
+            snapshot_id = stats.get("snapshot_id")
```
```suggestion
            snapshot_id = stats.get("snapshot-id")
```
i think this should be `snapshot-id` since before validator takes in json as input
https://github.com/apache/iceberg-python/blob/fa03e0814d8ed0a54be39e023f61e8163e37cdf3/pyiceberg/table/statistics.py#L32-L40

kevinjqliu replied:
could you add a test case for this (and possibly one for the else case too)? 

The current test only test the `StatisticsFile` instance branch
https://github.com/apache/iceberg-python/blob/fa03e0814d8ed0a54be39e023f61e8163e37cdf3/tests/table/test_init.py#L1370-L1381

kevinjqliu reviewed (approved):
LGTM!
